uefi: Add safe protocol wrapper for EFI_NVM_EXPRESS_PASS_THRU_PROTOCOL

seijikun opened the pull request:
**Update** [Latest API discussions](https://github.com/rust-osdev/uefi-rs/pull/1589#issuecomment-2791197828)

---

Implemented a safe API wrapper for `EFI_NVM_EXPRESS_PASS_THRU_PROTOCOL` [`52c78312-8edc-4233-98f2-1a1aa5e388a5`].
Added an integration-test for it.

~~+This contains a copy of the `AlignedBuffer` commit from https://github.com/rust-osdev/uefi-rs/pull/1589.+~~

This uses the iterator/mutable design I suggested here: https://github.com/rust-osdev/uefi-rs/pull/1589#issuecomment-2755354089 where the iter method only captures the protocol (`NvmePassThru`) immutably, but requires you to use the produced element (`NvmeNamespace`) in a mutable way for meaningful methods like `execute_command`.

## Checklist
- [x] Sensible git history (for example, squash "typo" or "fix" commits). See the [Rewriting History](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) guide for help.
- [x] Update the changelog (if necessary)

phip1611 commented on uefi-raw/src/protocol/nvme.rs:
```diff
@@ -48,7 +107,7 @@ pub struct NvmExpressPassThruCommandPacket {
     pub meta_data_buffer: *mut c_void,
     pub meta_data_length: u32,
-    pub queue_type: u8,
+    pub queue_type: NvmExpressQueueType,
```
nice, thanks for adding the specific hype here!

phip1611 commented on uefi-raw/src/protocol/nvme.rs:
```diff
@@ -30,8 +77,20 @@ pub struct NvmExpressCommand {
+
 /// This structure maps to the NVM Express specification Completion Queue Entry
-#[derive(Debug)]
+#[derive(Debug, Default)]
```
nit: why do we need Default here?

seijikun replied:
It's used in `NvmeNamespace::execute_command()`.
That struct is filled by the device, so it's never initialized with any meaningful values from the rust side.

phip1611 commented on uefi-test-runner/src/proto/nvme/pass_thru.rs:
```diff
@@ -0,0 +1,53 @@
+    assert!(try_find_drive());
+}
+
+fn try_find_drive() -> bool {
```
nit: `try_` functions return a `Result` by convention. Maybe `has_nvme_Device` is a better name?

phip1611 commented on uefi-test-runner/src/proto/nvme/pass_thru.rs:
```diff
@@ -0,0 +1,53 @@
+
+        let request = NvmeRequestBuilder::new(nvme_pt.io_align(), 0x06, NvmeQueueType::ADMIN)
+            .with_timeout(Duration::from_millis(500))
+            .with_cdw10(1) // we want info about controller
```
Not sure about NVME and how it works, but `.with_cdw10(1) /` seems a little magic to me.

Can we add high-level descriptions for that? Is this possible?

seijikun replied:
I added more documentation to clarify what these are ...

At least as far as I understood it:
The easiest way to think about them is as registers that get filled by a certain pattern (`calling convention`) to pass parameters to a called `C` function. Their interpretation is not fixed, it depends on the function being called.

phip1611 reviewed (approved):
LGTM, thanks for the high-quality contribution! As this is a bigger change and non trivially to review, I'd also like to wait for the feedback of Nicholas. So it may take some time (days-weeks) before we can get this merged.

seijikun commented on uefi/src/proto/nvme/pass_thru.rs:
```diff
@@ -0,0 +1,214 @@
+/// One protocol instance corresponds to one NVMe controller
+/// (which, most of the time, corresponds to one SSD).
+///
+/// This API offers a safe and convenient, yet still low-level interface to ATA/NVMe/SCSI devices.
```
whoops. Too fast ;)

seijikun commented:
I just:
- rebased this PR onto master
- attempted to split it up into smaller commits as good as I can, in an attempt to ease the review process for you
- added some more non-docs comments with some reasoning and code explanations

phip1611 commented on uefi/src/proto/nvme/mod.rs:
```diff
@@ -0,0 +1,285 @@
+    /// # Returns
+    /// An instance of [`NvmeRequestBuilder`] for further configuration.
+    #[must_use]
+    pub fn new(io_align: u32, opcode: u8, queue_type: NvmeQueueType) -> Self {
```
I think we could add an assert to check that `io_align: u32` is a power of 2, but that is not urgent to get this merged

phip1611 reviewed (approved):
LGTM, thanks for your patience! I'm not very experienced with NVMe but these abstractions look good. A clear value-add to this crate
